sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -40,23 +40,47 @@ class EquivalentExpressions {
     override def hashCode: Int = e.semanticHash()
   }
 
+  /**
+   * Wrapper around an Expression that provides structural semantic equality.
+   */
+  case class StructuralExpr(e: Expression) {
+    def normalized(expr: Expression): Expression = {
+      expr.transformUp {
+        case b: ParameterizedBoundReference =>
+          b.copy(parameter = "")
+      }
+    }
+    override def equals(o: Any): Boolean = o match {
+      case other: StructuralExpr =>
+        normalized(e).semanticEquals(normalized(other.e))
+      case _ => false
+    }
+
+    override def hashCode: Int = normalized(e).semanticHash()
+  }
+
+  type EquivalenceMap = mutable.HashMap[Expr, mutable.ArrayBuffer[Expression]]
+
   // For each expression, the set of equivalent expressions.
   private val equivalenceMap = mutable.HashMap.empty[Expr, mutable.ArrayBuffer[Expression]]
 
+  // For each expression, the set of structurally equivalent expressions.
+  private val structEquivalenceMap = mutable.HashMap.empty[StructuralExpr, EquivalenceMap]
```

Comment:
Naturally thinking from the comment, I thought the type's `mutable.HashMap.empty[StructuralExpr, mutable.ArrayBuffer[Expression]]`. Is  the value `EquivalenceMap` because you reuse the existing logic?

Reply:
Among expressions with same structure, there are still different sub-set of expressions which are semantically different to each others. I need the EquivalenceMap here to distinguish them.

Reply:
Yea, it looks ok. So, could you update the comment to describe more?